docs/source/python/integration/python_r.rst

```diff
@@ -0,0 +1,312 @@
+necessary to import an Arrow Array in R from the C Data interface.
+
+That work will be done by the ``addthree_cdata`` function which invokes the
+``addthree`` function once the Array is imported.
+
+Our ``addthree.R`` will thus have both the ``addthree_cdata`` and the 
+``addthree`` functions:
+
+.. code-block:: R
+
+    library(arrow)
+
+    addthree_cdata <- function(array_ptr_s, schema_ptr_s) {
+        array_ptr <- as.numeric(array_ptr_s)
+        schema_ptr <- as.numeric(schema_ptr_s)
+
+        a <- Array$import_from_c(array_ptr, schema_ptr)
+
+        return(addthree(a))
+    }
+
+    addthree <- function(arr) {
+        return(arr + 3)
+    }
+
+We can now provide to R the array and its schema from Python through the
+``array_ptr_s`` and ``schema_ptr_s`` arguments so that R can build back
+an ``Array`` from them and then invoke ``addthree`` with the array.
+
+Invoking ``addthree_cdata`` from Python involves building the Array we
+want to pass to ``R``, exporting it to the C Data interface and then
+passing the exported references to the ``R`` function.
+
+Our ``addthree.py`` will thus become:
+
+.. code-block:: python
+
+    # Get a reference to the addthree_cdata R function
+    import rpy2.robjects as robjects
+    r_source = robjects.r["source"]
+    r_source("addthree.R")
+    addthree_cdata = robjects.r["addthree_cdata"]
+
+    # Create the pyarrow array we want to pass to R
+    import pyarrow
+    array = pyarrow.array((1, 2, 3))
+
+    # Import the pyarrow module that provides access to the C Data interface
+    from pyarrow.cffi import ffi as arrow_c
+
+    # Allocate structures where we will export the Array data 
+    # and the Array schema. They will be released when we exit the with block.
+    with arrow_c.new("struct ArrowArray*") as c_array, \
+         arrow_c.new("struct ArrowSchema*") as c_schema:
+        # Get the references to the C Data structures.
+        c_array_ptr = int(arrow_c.cast("uintptr_t", c_array))
+        c_schema_ptr = int(arrow_c.cast("uintptr_t", c_schema))
+
+        # Export the Array and its schema to the C Data structures.
+        array._export_to_c(c_array_ptr)
+        array.type._export_to_c(c_schema_ptr)
+
+        # Invoke the R addthree_cdata function passing the references
+        # to the array and schema C Data structures. 
+        # Those references are passed as strings as R doesn't have
+        # native support for 64bit integers, so the integers are
+        # converted to their string representation for R to convert it back.
+        r_result_array = addthree_cdata(str(c_array_ptr), str(c_schema_ptr))
+
+        # r_result will be an Environment variable that contains the
+        # arrow Array built from R as the return value of addthree.
+        # To make it available as a Python pyarrow array we need to export
+        # it as a C Data structure invoking the Array$export_to_c R method
+        r_result_array["export_to_c"](float(c_array_ptr), float(c_schema_ptr))
```

Comment:
Ouch, why does this cast to `float`? Floats cannot represent all 64-bit numbers faithfully.

Reply:
That's sadly a side effect of interfacing to R. Integers in R are 32-bits and thus if you try to pass around the pointers as integers they will overflow. You get a `integer 140411646274592 does not fit '32-bit int'` kind of error.

For the `addthree_cdata` function (as it was implemented by myself) I passed the pointers as strings, so they would represent the exact value, but for `Array$export_to_c` as it's a R function expecting a number I couldn't provide a string, I had to pass a `REAL` as expected by the function. Thus I had to go for a `float` to have `rpy2` convert it to a `REAL`.

I guess an alternative might be to get a reference to the `as.numeric` R function and invoke `as.numeric(str(c_array_ptr))` that would force the conversion to an R numeric from the string in R itself so we would avoid going through a Python float.

Reply:
What does it change to call `as.numeric`? You would still get a floating-point number with not enough precision, right?

Reply:
Changed the example to use `as.numeric` instead of `float`

Reply:
@pitrou Well, I tried with `bit64` `integer64` type, but it didn't work, it seems to only accept a REAL, if I pass a `integer64` to `export_to_c` or `import_from_c` I get a segfault. Someone with more R knowledge might want to chime in. @jonkeane @thisisnic